### PR TITLE
TimeSeries samples now Truncated, not Accumulated

### LINKED ISSUE #5658: stability: crash on checksum mismatch
scenario:
* 4 nodes running with 4 photos apps. 2 died overnight of OOM and #5621.
* stopped all remaining nodes
* upgraded from sha a52f363 to de433e9
* restarted all nodes (no load yet)

after about 6 minutes, two nodes crashed with:
```
F160329 13:34:56.024740 storage/replica_command.go:1617  replica: range=1 [/Min-/System/"tsd\x12cr.node.txn.durations-1m-p50\x00\x01\x89\xf8\x06/a2"), checksum mismatch: e = 44a2ac0be2ce7548f90f9ab2730424812aaa49463a8ae4d69441f61494ba289b8c2418128b48d25984b5cad3c96c8cdd5519366b8de949b7154ca8fc1710835b, v = d7a555600dc25fb1ec0ebd8cde1717f229d872d09d6dafc5f4328e302fe1facb5fccc90ac4a1ae2474421661c7881dabd1cc1ec6e476c9d828bd1dcf716ed72f
goroutine 51 [running]:
github.com/cockroachdb/cockroach/util/log.getStacks(0xece8c7b00, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/util/log/clog.go:858 +0xb8
github.com/cockroachdb/cockroach/util/log.(*loggingT).outputLogEntry(0x2638060, 0x3, 0x203986a, 0x1a, 0x651, 0xc821886f00, 0x17b)
        /go/src/github.com/cockroachdb/cockroach/util/log/clog.go:722 +0x178
github.com/cockroachdb/cockroach/util/log.addStructured(0x0, 0x0, 0xe00000003, 0x2, 0x1a0b8e0, 0x2e, 0xc82210fb38, 0x3, 0x3)
        /go/src/github.com/cockroachdb/cockroach/util/log/structured.go:57 +0xfc
github.com/cockroachdb/cockroach/util/log.logDepth(0x0, 0x0, 0x1, 0xc800000003, 0x1a0b8e0, 0x2e, 0xc82210fb38, 0x3, 0x3)
        /go/src/github.com/cockroachdb/cockroach/util/log/log.go:73 +0x7b
github.com/cockroachdb/cockroach/util/log.Fatalf(0x1a0b8e0, 0x2e, 0xc82210fb38, 0x3, 0x3)
        /go/src/github.com/cockroachdb/cockroach/util/log/log.go:179 +0x66
github.com/cockroachdb/cockroach/storage.(*Replica).VerifyChecksum(0xc8202b20e0, 0x7f981efb6fa8, 0xc823c3de90, 0xc822110fe8, 0x144053de6628126d, 0x0, 0x0, 0x0, 0x1, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/storage/replica_command.go:1617 +0x617
github.com/cockroachdb/cockroach/storage.(*Replica).executeCmd(0xc8202b20e0, 0x7f981f0027c0, 0xc823c3de60, 0xc823c39a00, 0x8, 0x0, 0x7f981efb6fa8, 0xc823c3de90, 0xc822110fe8, 0x144053de6628126d, ...)
        /go/src/github.com/cockroachdb/cockroach/storage/replica_command.go:168 +0x1b10
github.com/cockroachdb/cockroach/storage.(*Replica).executeBatch(0xc8202b20e0, 0x7f981f0027c0, 0xc823c3de60, 0xc823c39a00, 0x8, 0x7f981efb6fa8, 0xc823c3de90, 0xc822110fe8, 0x144053de6628126d, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1694 +0x237
github.com/cockroachdb/cockroach/storage.(*Replica).applyRaftCommandInBatch(0xc8202b20e0, 0x7f981f0027c0, 0xc823c3de60, 0xc823c39a00, 0x8, 0x300000003, 0x6, 0x144053de6628126d, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1574 +0x495
github.com/cockroachdb/cockroach/storage.(*Replica).applyRaftCommand(0xc8202b20e0, 0xc823c39a00, 0x8, 0x7f981f0027c0, 0xc823c3de60, 0x21676, 0x300000003, 0x6, 0x144053de6628126d, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1470 +0x4a7
github.com/cockroachdb/cockroach/storage.(*Replica).processRaftCommand(0xc8202b20e0, 0xc823c39a00, 0x8, 0x21676, 0x1, 0x300000003, 0x6, 0x144053de6628126d, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1435 +0x2be
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc8202b20e0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1299 +0xa49
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
        /go/src/github.com/cockroachdb/cockroach/storage/store.go:1861 +0x35a
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202b71f0, 0xc820954910)
        /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
        /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
```

The second node died with the almost the same message, just different `v`:
```
F160329 13:34:53.933187 storage/replica_command.go:1617  replica: range=1 [/Min-/System/"tsd\x12cr.node.txn.durations-1m-p50\x00\x01\x89\xf8\x06/a2"), checksum mismatch: e = 44a2ac0be2ce7548f90f9ab2730424812aaa49463a8ae4d69441f61494ba289b8c2418128b48d25984b5cad3c96c8cdd5519366b8de949b7154ca8fc1710835b, v = 59518e3341bab7d6bbf348136faab8bcc951be2bc4b416a849139d3ee5a53683f7729e3f1f2555b4b66c95001930525fd39563ddd3f3bd7aa4a69cce8a8c4a92
```

### PR DESCRIPTION
**Pull request by mrtracy**
This commit modifies the way our time series system handles multiple values
in the same "sample period". Previously, the engine would accumulate the values
from the individual samples, maintaining a "sum", "count", "max" and "min".

However, there were a variety of consistency issues when this interacted with
the concept of "partial merges":
- Our ability to provide protection from replayed raft commands was different,
  depending on how partial merges were applied.
- The accumulation of floating point values is not associative, and the order
  could be different depending on how partial merges are applied.

Most troublingly, this could result on inconsistencies between raft replicas -
partial merges may occur on one replica independently of the others, resulting
in differences in the final merged result on disk. This is unacceptable.

Unfortunately, support for "partial merges" is relatively important for
performance reasons, and we are hesitant to remove it.

To fix all problems associated with this, we have decided to remove sample
accumulation. As an alternative, when multiple samples exist for the same sample
offset, they are discarded except for the most recently merged. This is
considered an acceptable compromise for all time series maintained by cockroach.

Additional Changes:
- Updated ts/doc.go to better reflect the current state of the /ts package.
- Added a test designed by @bdarnell to test replay protection in the scope of
  compactions.
- Removed generalized "replay protection" in merge (which did not work) and
  added an advisory for any future programmers looking to add a mergeable type.

Fixes #5658

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5905)

### CODE REVIEW

File: storage/engine/rocksdb/db.cc [hunk 1]
```diff
@@ -645,17 +649,13 @@ bool MergeValues(cockroach::storage::engine::MVCCMetadata *left,
...
+
+    // Replay Advisory: Because merge commands pass through raft, it is possible
+    // for merging values to be "replayed". Currently, the only actual use of
+    // the merge system is for time series data, which is safe against replay;
+    // however, this property is not general for all potential mergeable types.
+    // If a future need arises to merge another type of data, replay protection
+    // will likely need to be a consideration.
     if (left->has_merge_timestamp() && right.has_merge_timestamp()) {
       if (left->merge_timestamp().wall_time() == right.merge_timestamp().wall_time() &&
           left->merge_timestamp().logical() == right.merge_timestamp().logical()) {
```
Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: Didn't you want to disable the replay protection here?

Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: Ah, never mind. You're saying that merge replays are now safe due to the way merges are processed.

Comment (mrtracy) [inline storage/engine/rocksdb/db.cc]: No, you were right the first time, I wanted to remove this.  I had a weird merge at the very end, didn't delete enough.

**Comment by tamird** [comment]
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/engine/mvcc_test.go, line 3246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEheW1ggau2KxLB1W-i:-KEheW1hnC2xwzmF1cBE:-1030970629) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/storage/engine/mvcc_test.go#L3246)):*
slightly more type-safe is proto.Equal (reflect.DeepEqual can fail spuriously when one argument is a pointer and the other is not)

---

*[storage/engine/mvcc_test.go, line 3248 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEheihiQcCgcRxnxGfG:-KEheihiQcCgcRxnxGfH:-626323922) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/storage/engine/mvcc_test.go#L3248)):*
this is unmarshalling into a null pointer; no bueno. make `{first,second}TS` values?

---

*[ts/doc.go, line 19 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhfOysqfLBjFoMeFLG:-KEhfOysqfLBjFoMeFLH:1538858389) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/ts/doc.go#L19)):*
s/DV/DB/

---

*[ts/doc.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhfUaJFMLTBvDyRLPi:-KEhfUaJFMLTBvDyRLPj:1959441199) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/ts/doc.go#L26)):*
s/these/those/

---

*[ts/doc.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhft2n6EWXf_w9Boez:-KEhft2n6EWXf_w9Bof-:-687702427) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/ts/doc.go#L77)):*
s/csae/case/

---

*[ts/doc.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhfzjiHRVTjWJclWYs:-KEhfzjiHRVTjWJclWYt:-1844565859) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/ts/doc.go#L89)):*
too many spaces

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905)*

**Comment by vivekmenezes** [comment]
can you update internal.proto to say

1. max and min are no longer populated and are here for backward compatibility
2. count will always be 1 in the future and is here for backward compatibility.

This will allow us to remove these values in the future

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/engine/mvcc_test.go, line 3216 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEiFR_QIHGxn1jEsc1l:-KEiFR_QIHGxn1jEsc1m:-1280936503) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/storage/engine/mvcc_test.go#L3216)):*
remove

---

*[storage/engine/rocksdb/db.cc, line 661 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhStm2sydGRkrOHER_-r1-661:-KEiEdtoO2xmkxVIghc5:428828946) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/storage/engine/rocksdb/db.cc#L661)):*
yes please delete this code

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-:-KEiIhsOeIrMwTVhs67g:396460319)*

**Comment by mrtracy** [comment]
Comment added to internal.proto.

---

Review status: 5 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/engine/mvcc_test.go, line 3216 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEiFR_QIHGxn1jEsc1l:-KEiumw8whZ6pYRYd0I9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/storage/engine/mvcc_test.go#L3216)):*
Done.

---

*[storage/engine/mvcc_test.go, line 3246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEheW1ggau2KxLB1W-i:-KEiuEI0iR7xWBXgwcca:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/storage/engine/mvcc_test.go#L3246)):*
Done.

---

*[storage/engine/mvcc_test.go, line 3248 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEheihiQcCgcRxnxGfG:-KEiuFF6zBsfapacWwjX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/storage/engine/mvcc_test.go#L3248)):*
Done.

---

*[storage/engine/rocksdb/db.cc, line 661 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhStm2sydGRkrOHER_-r1-661:-KEiuGafxipSrfM2MIHA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/storage/engine/rocksdb/db.cc#L661)):*
Done.

---

*[ts/doc.go, line 19 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhfOysqfLBjFoMeFLG:-KEiuUPDz7W7wm_C5Lad:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/ts/doc.go#L19)):*
Done.

---

*[ts/doc.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhfUaJFMLTBvDyRLPi:-KEiuaWoo_dPkbc0ZxAb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/ts/doc.go#L26)):*
Done.

---

*[ts/doc.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhft2n6EWXf_w9Boez:-KEiuiB1sruA8nWy_At7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/ts/doc.go#L77)):*
Done.

---

*[ts/doc.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEhfzjiHRVTjWJclWYs:-KEiulDGcxZOy46zDFSY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c3f4b99dc067d5345760d2fc7bbde71164ef4/ts/doc.go#L89)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-:-KEiv_UI02FlC0j8rG3I:-1163530536)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[roachpb/internal.proto, line 87 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEkIkDgZz_3fvuQg4-O:-KEkIkDgZz_3fvuQg4-P:-488633117) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e2e7fe70e83921d67439a618b9dbedc4495da00/roachpb/internal.proto#L87)):*
you should actually rename these to indicate their deprecation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-:-KEkIq7R-Sg2oGfFr8MB:1417088721)*

**Comment by petermattis** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/engine/rocksdb.go, line 307 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEkfhCJ7qq2r4pL2Um2:-KEkfhCJ7qq2r4pL2Um3:-795434221) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e2e7fe70e83921d67439a618b9dbedc4495da00/storage/engine/rocksdb.go#L307)):*
Using compaction to test the merge logic feels indirect. On the plus side, it is exercising the actual code paths that will be used in production, but I'm wondering if it would be better to enhance `goMerge` (which is only used for unit tests) to specify whether a full or partial merge is desired.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-:-KEkgRXa_zcX96c8GcBc:1417088721)*

**Comment by vivekmenezes** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[roachpb/internal.proto, line 87 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEkIkDgZz_3fvuQg4-O:-KEkuXy15w0jmEbNtuyk:767694897) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e2e7fe70e83921d67439a618b9dbedc4495da00/roachpb/internal.proto#L87)):*
I think we can leave them as is because they are still being read. Perhaps we should add the DELETED_ prefix to them after say 3 months and remove them from the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-:-KEkv6FPlYF2kC_Bu1lc:1417088721)*

**Comment by bdarnell** [comment]
Review status: 9 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/engine/rocksdb.go, line 307 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905#-KEkfhCJ7qq2r4pL2Um2:-KElzNGeOJp65vB2Vsl6:1589562090) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e2e7fe70e83921d67439a618b9dbedc4495da00/storage/engine/rocksdb.go#L307)):*
Yeah, this does feel fragile. I verified by hand that this caused partial merges to occur but that's kind of an implementation detail and the test isn't confirming that that the compactions are doing anything at all.  Adding a test using `goMerge` with a new flag sounds like a good idea.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5905)*

**Pull request merged**